plugins/gcloud/lib/samson_gcloud/image_tagger.rb

```diff
@@ -15,11 +17,20 @@ def tag(deploy)
           next unless digest.match?(/(^|\/|\.)gcr.io\//) # gcr.io or https://gcr.io or region like asia.gcr.io
           base = digest.split('@').first
 
-          if deploy.stage.production?
-            tag_image('production', deploy, base, digest)
+          tags = []
+          stage = deploy.stage
+          if stage.production? && !stage.no_code_deployed?
+            tags << PRODUCTION_TAG
+          end
+
+          if env_permalink = stage.environments&.first&.permalink
+            tags << "env-#{env_permalink}"
           end
 
-          tag_image(deploy.stage.permalink, deploy, base, digest)
+          tags << stage.permalink
+
+          # From docker distro: https://tinyurl.com/ycgh7qsp
+          tags.grep(/^[\w][\w.-]{0,127}$/).each { |tag| tag_image(tag, deploy, base, digest) }
```

Comment:
`.uniq` since the permalink can be production too

Reply:
Would we want to avoid this collision completely by prefixing the stage permalink tags?